Refactor SuggestionsList to use hooks and change class to function component

amustaque97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Introduced functional component. In the component I'm using useState, useEffect.
Renamed a few variables to get rid of syntax errors/conflicts. Tested the component and
it is working as expected.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. In wp admin go to tags page.
<img width="198" alt="image" src="https://user-images.githubusercontent.com/21127788/143674372-3c778ecd-6493-4141-8f69-02262ed63422.png">
2. Create multiple tags
<img width="1445" alt="image" src="https://user-images.githubusercontent.com/21127788/143674405-55f34f53-3069-40c0-9847-55b084da50e5.png">

3. Edit or create new post.

4. In the settings, try to add new tags.
5. It will show  suggestion list drop-down
<img width="280" alt="image" src="https://user-images.githubusercontent.com/21127788/143674575-a65eda63-aa84-4195-a55d-7f08b9e9ee0b.png">

6. Select few tags 
7. Publish the post.
## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Code Quality / Refactor
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

amustaque97 committed:
Refactor SuggestionsList to use hooks and change class to function component

Introduced functional component. In the component I'm using useState, useEffect.
Renamed a few variables to get rid of syntax errors. Tested the component and
it is working as expected.

amustaque97 committed:
fix scrollIntoView import issue and fix component test case

There was an error saying scrollIntoView function not found. I was not
importing it properly now in this commit I have renamed scrollIntoView function
from dom-scroll-into-view package is renamed as scrollView. I had to do this because
there is already props with the same name ie scrollIntoView. Component test case was
failing because of issue on the same lines that is now fixed.

There was another error related to not using act from react-dom/test-utils package in
test case. I have fixed it. All component test case are passing now.

amustaque97 committed:
Merge remote-tracking branch 'origin/trunk' into refactor/suggestion-list-component

amustaque97 committed:
Merge remote-tracking branch 'origin/trunk' into refactor/suggestion-list-component

ciampo commented on packages/components/src/form-token-field/test/index.js:
```diff
@@ -262,7 +262,9 @@ describe( 'FormTokenField', () => {
 			// before sending a hover event, we need to wait for
 			// SuggestionList#_scrollingIntoView to become false
-			jest.advanceTimersByTime( 100 );
+			act( () => {
+				jest.advanceTimersByTime( 100 );
+			} );
```
Why was this change necessary?

amustaque97 replied:
Half of the month is gone. Don't remember exact details 😅 . Sorry about that. But yeah test case was failing and after going to test failure. I learned more about act and I was able to fix it. I come to know about `act` from test failure error message as shown in the below image.
<img width="1370" alt="Screenshot 2021-12-16 at 12 24 17 AM" src="https://user-images.githubusercontent.com/21127788/146247883-8933fc89-4fee-4121-bad8-e219075214b5.png">

ciampo reviewed (commented):
Thank you @amustaque97 for working on this ✌️ 

I'm going to give this a proper look later (either by end of day or on Friday), but I gave a quick look and code changes overall look good to me.

ciampo commented:
Hey @amustaque97, sorry if I haven't managed to take a proper look at this yet, but I had a few unforeseen PRs that took precedence. I will probably be able to do to in the new year, since I'm going to be AFK in the next couple of weeks. Hope that's ok!

ZebulanStanphill commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -2,157 +2,147 @@
+	selectedIndex,
+	scrollIntoView,
+	match = '',
+	onHover = () => {},
+	onSelect = () => {},
```
A slight optimization could be made by using lodash's `noop` here to reduce the number of functions being created, e.g.:

```js
import { map, noop } from 'lodash'
// ...
	onHover = noop,
	onSelect = noop,
```

ciampo replied:
Premise: nit

Just wanted to add that another approach is to leave those function props as `undefined` by default, and use the optional chaining syntax when invoking them, e.g.

```js
const handHover = () => {
  // do stuff

  onHover?.();
}
```

This has the the advantage of relying solely on native APIs/syntax (plus it avoids initialising those values to their default value). But this is more of a personal preference / micro-optimisation, so we can keep the code as-is.

ZebulanStanphill replied:
I agree that using optional chaining and an `undefined` default is better than using lodash's `noop`. Thanks for pointing that out.

amustaque97 replied:
Thanks for the helping me here 😅 I would prefer to keep things simple. Your comment will help others as well to understand of writing this way instead of `noop`

ciampo replied:
Just seen your last commit — I believe that the alternative to using `noop`  that we agreed on should look more like this:

```diff
diff --git a/packages/components/src/form-token-field/suggestions-list.js b/packages/components/src/form-token-field/suggestions-list.js
index 3794448014..b7eee38e95 100644
--- a/packages/components/src/form-token-field/suggestions-list.js
+++ b/packages/components/src/form-token-field/suggestions-list.js
@@ -22,8 +22,8 @@ function SuggestionsList( {
 	selectedIndex,
 	scrollIntoView,
 	match = '',
-	onHover = () => {},
-	onSelect = () => {},
+	onHover,
+	onSelect,
 	suggestions = emptyList,
 	displayTransform,
 	instanceId,
@@ -55,14 +55,14 @@ function SuggestionsList( {
 	const handleHover = ( suggestion ) => {
 		return () => {
 			if ( ! scrollingIntoView ) {
-				onHover( suggestion );
+				onHover?.( suggestion );
 			}
 		};
 	};
 
 	const handleClick = ( suggestion ) => {
 		return () => {
-			onSelect( suggestion );
+			onSelect?.( suggestion );
 		};
 	};
 
```

ZebulanStanphill commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -2,157 +2,147 @@
+	match = '',
+	onHover = () => {},
+	onSelect = () => {},
+	suggestions = Object.freeze( [] ),
```
`Object.freeze( [] )` could be moved out of this function into a `const` to avoid recreating it on every render.

ZebulanStanphill commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -2,157 +2,147 @@
+	};
 
-	handleMouseDown( e ) {
+	const handleMouseDown = ( e ) => {
```
This function could be moved to the top level of the file to avoid recreating it on every render.

amustaque97 committed:
Address review comments

- Add `noop` to reduce number of functions creation
- Move  emptyList to a const then use it
- Move `handleMouseDown` to top of the file to avoid recreating it on every render

amustaque97 commented:
All review comments looks logical to me. I have addressed review comments. Thank You for the review. :smile:

ciampo commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -1,158 +1,150 @@
+	instanceId,
+	setTimeout,
+} ) {
+	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
```
Shouldn't the default value of `scrollingIntoView` be `false` ?

amustaque97 replied:
Oops! you're right 💯 

Correct me If I'm wrong here, I checked variable here https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/form-token-field/index.js#L50
and based on that I'm agreeing its value should be false.

ciampo replied:
My reasoning was that, prior to your refactor, `this.scrollingIntoView` was not initialised to anything — and therefore it held a value of `undefined`. Since we're refactoring the code to `useState` and we want that variable to always hold a boolean value, then we should use `false` (since `undefined` is a falsy value)

Hope this is clear enough!

ciampo commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -1,158 +1,150 @@
+	setTimeout,
+} ) {
+	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
+	const [ list, setList ] = useState();
```
Instead of using state and a custom ref callback to store a `ref` to the `<ul>` element, we can use the [custom `useRefEffect` hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts):

<details>
<summary>Click to expand suggested changes</summary>

```diff
diff --git a/packages/components/src/form-token-field/suggestions-list.js b/packages/components/src/form-token-field/suggestions-list.js
index 397a33863e..1c0d770af6 100644
--- a/packages/components/src/form-token-field/suggestions-list.js
+++ b/packages/components/src/form-token-field/suggestions-list.js
@@ -8,8 +8,8 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { useEffect, useState } from '@wordpress/element';
-import { withSafeTimeout } from '@wordpress/compose';
+import { useState } from '@wordpress/element';
+import { withSafeTimeout, useRefEffect } from '@wordpress/compose';
 
 const emptyList = Object.freeze( [] );
 
@@ -30,30 +30,27 @@ function SuggestionsList( {
 	setTimeout,
 } ) {
 	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
-	const [ list, setList ] = useState();
+	const listRef = useRefEffect(
+		( listNode ) => {
+			// only have to worry about scrolling selected suggestion into view
+			// when already expanded
+			if (
+				selectedIndex > -1 &&
+				scrollIntoView &&
+				listNode.children[ selectedIndex ]
+			) {
+				setScrollingIntoView( true );
+				scrollView( listNode.children[ selectedIndex ], listNode, {
+					onlyScrollIfNeeded: true,
+				} );
 
-	useEffect( () => {
-		// only have to worry about scrolling selected suggestion into view
-		// when already expanded
-		if (
-			selectedIndex > -1 &&
-			scrollIntoView &&
-			list.children[ selectedIndex ]
-		) {
-			setScrollingIntoView( true );
-			scrollView( list.children[ selectedIndex ], list, {
-				onlyScrollIfNeeded: true,
-			} );
-
-			setTimeout( () => {
-				setScrollingIntoView( false );
-			}, 100 );
-		}
-	}, [ selectedIndex, scrollIntoView, list ] );
-
-	const bindList = ( ref ) => {
-		setList( ref );
-	};
+				setTimeout( () => {
+					setScrollingIntoView( false );
+				}, 100 );
+			}
+		},
+		[ selectedIndex, scrollIntoView ]
+	);
 
 	const handleHover = ( suggestion ) => {
 		return () => {
@@ -98,7 +95,7 @@ function SuggestionsList( {
 	// TODO does this still apply now that it's a <ul> and not a <div>?
 	return (
 		<ul
-			ref={ bindList }
+			ref={ listRef }
 			className="components-form-token-field__suggestions-list"
 			id={ `components-form-token-suggestions-${ instanceId }` }
 			role="listbox"
```
</details>

This could potentially help to fix the [e2e test error](https://github.com/WordPress/gutenberg/runs/4575862072?check_suite_focus=true), since the callback in `useRefEffect` is invoked only if `listNode` isn't falsy.

amustaque97 replied:
Can you please help me in understanding where it was wrong in my code? Just curious

ciampo replied:
There wasn't anything _wrong_ in your code — you were just refactoring the component from class to function — my suggestion here is more of an improvement on top of your changes. I believe the code is cleaner (and potentially causes less re-renders)

Here's the reasoning that brought me to suggest `useRefEffect`:

- While reviewing your PR, I was having a look at the `list` state, and noticed that it was used to hold a `ref`.
- Now, we could have used `useRef` instead, but then I noticed that `list` was used as a dependency of the `useEffect` hook — switching to `useRef` would meant that changes to `list` wouldn't trigger the `useEffect` hook.
- The proper way to refactor the code, then, would be to rewrite `list` to use `useCallback` to hold the `ref`. The code in `useEffect` (and its dependencies) could have moved to the body of `useCallback` too
- Finally, I realised that the pattern that I just described is implemented in a custom hook called [`useRefEffect`](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts)

amustaque97 replied:
Thanks for the brief explanation.

ciampo commented on packages/components/src/form-token-field/suggestions-list.js:
```diff
@@ -1,158 +1,150 @@
+		>
+			{ map( suggestions, ( suggestion, index ) => {
+				const matchText = computeSuggestionMatch( suggestion );
+				const classeName = classnames(
```
Nit: typo

```suggestion
				const className = classnames(
```

(we should then update any reference to this variable)

amustaque97 replied:
Apologies for this 😅 

I have tested again my changes. Working as expected. :smile:

ciampo reviewed (commented):
Thank you for working on this, @amustaque97 !

I took some time, had a proper look at the code and left some comments

amustaque97 committed:
refactor `noop`, set default value `scrollingIntoView` false and use `useRefEffect`

ciampo reviewed (commented):
Thank you for addressing the feedback! Not crucial, but in the future it would be even better if you could split different sets of changes over separate commits

I just left a minor comment, otherwise the code changes LGTM and tests well.

Could you also add an entry to the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) ?

amustaque97 committed:
Merge branch 'trunk' into refactor/suggestion-list-component

amustaque97 committed:
add changelog entry

amustaque97 committed:
set `onHover` and `onSelect` default value undefined and use optional chaining
